src/neuralnet/layer.cc

```diff
@@ -160,6 +160,200 @@ void DropoutLayer::ComputeGradient(const vector<SLayer>& srclayers)  {
   Tensor<cpu, 1> gsrc(gsrcblob->mutable_cpu_data(), Shape1(gsrcblob->count()));
   gsrc=grad*mask;
 }
+/**************** Implementation for DBMBottomLayer********************/
+void DBMBottomLayer::Setup(const LayerProto& proto,
+      const vector<SLayer>& srclayers) {
+  CHECK_EQ(srclayers.size(), 1);
+  kPhase = true;
+  const auto& possrc = srclayers[0]->data(this);
+  is_first_iteration_bottom = true;
+  batchsize_ = possrc.shape()[0];
+  neg_batchsize_ = batchsize_;
+  /*gibbs sampling size and input have the same size*/
+  vdim_ = possrc.count()/batchsize_;
+  hdim_ = proto.dbmbottom_conf().num_output();
+  data_.Reshape(vector<int>{batchsize_, hdim_});
+  hidden_data_.Reshape(vector<int>{neg_batchsize_, hdim_});
+  scale_ = static_cast<float> (1.0f);
+  negsrc_.Reshape(vector<int>{neg_batchsize_, vdim_});
+  Factory<Param>* factory = Singleton<Factory<Param>>::Instance();
+  weight_ = shared_ptr<Param>(factory->Create("Param"));
+  bias_ = shared_ptr<Param>(factory->Create("Param"));
+  weight_->Setup(proto.param(0), vector<int>{vdim_, hdim_}, vdim_*hdim_);
+  bias_->Setup(proto.param(1), vector<int>{vdim_}, 0);
+  srand((unsigned)time(NULL));
+}
+void DBMBottomLayer::SetupAfterPartition(const LayerProto& proto,
+      const vector<int> &shape,
+      const vector<SLayer>& srclayers) {
+}
+
+void DBMBottomLayer::ComputeFeature(Phase phase, const vector<SLayer>& srclayers) {
```

Comment:
can we compress the code by first define some pointers for neg phase or pos phase, and them use these pointers to create tensors and do the dot, sigmoid operations.